### PR TITLE
allow passing options to duplication engine's babylon

### LINKED ISSUE #128: Configurable Babylon plugins for JS
Right now, the selection of plugins passed to Babylon for parsing JavaScript files [appears to be hard-coded](https://github.com/codeclimate/codeclimate-duplication/blob/master/lib/cc/engine/analyzers/javascript/parser.js#L22). A one-size-fits-all approach doesn't seem ideal here, though — for instance, in our codebase we don't use JSX, but we do have a number of `async` functions, which are rejected as invalid syntax.

It would be great if the Babylon plugins could be set up as part of the configuration for the duplication engine for a given codebase.

### PR DESCRIPTION
**Pull request by rcsole**
closes #128

### CODE REVIEW

**Comment by rcsole** [comment]
Hi! It seems like current master also breaks in that one test? Does not
seem related to anything I've messed with.

**Comment by maxjacobson** [comment]
Hi @rcsole, thanks for making this PR.

When running the tests locally, are you using `make test`, which builds a Docker image and runs the tests inside a container?

When I run tests this way, the tests pass on master, but fail on your branch in the same way as on Circle CI.

**Comment by rcsole** [comment]
Oh, hey I was running them as `make citest` to mimic CircleCI. My bad!

**Comment by rcsole** [comment]
@maxjacobson should be fixed now 🤗 🤗 🤗

**Comment by maxjacobson** [comment]
Sweet!

What do you think updating the tests to cover the code paths of when this configuration is used vs not used?

**Comment by rcsole** [comment]
Sure! I won't have time today, but I might sometime this weekend.

**Comment by rcsole** [comment]
Hi @maxjacobson do you have a test coverage tool that I can run
locally? I'm struggling a little bit 😕

**Comment by maxjacobson** [comment]
@rcsole Sorry about that.

We do, although it looks like we haven't enabled it for this repo

This repo contains duplication analysis across a handful of languages, an our test suite is in Ruby. The relevant test, which you can extend is here: https://github.com/codeclimate/codeclimate-duplication/blob/57e977ba17f302ccbe7a4746d7d62cebb5f4198b/spec/cc/engine/analyzers/javascript/main_spec.rb

I'm sympathetic if you're more of a JavaScript developer and not necessarily familiar with Ruby. Would you like help with that?

**Comment by maxjacobson** [comment]
(I see you have enabled the new GitHub feature which allows me to push to your branch, but would like to ask permission first!)

> Add more commits by pushing to the master branch on rcsole/codeclimate-duplication.

**Comment by rcsole** [comment]
Go ahead!

**Comment by maxjacobson** [comment]
Oh sure. I'm fairly busy this week, so if you'd like to take a crack at it that would be great.

In terms of coverage, I'm mainly interested in covering the new behavior introduced in this PR. Specifically: when this new configuration is provided, and `flow` is added, does it correctly find duplication issues in code which uses the "flow" syntax?

File: README.md [hunk 1]
```diff
@@ -38,6 +38,26 @@ engines:
       - javascript
 ```
 
+#### JavaScript
+
+The engine uses `babylon` to parse JS source code, you can choose which plugins
+to be included by flow with the following example (default is only `jsx`):
```
Comment (maxjacobson) [inline README.md]: should this say "to be included by babylon"?

**Comment by dblandin** [comment]
Hi! Is there still interest in finishing up this PR? How much work is left to push it over the line?

**Comment by rcsole** [comment]
Hi @dblandin, I think what's left is the tests, but I, unfortunately, don't have time to finish it up. I'm sorry this has been sitting here so long.

**Comment by dblandin** [comment]
Going to close this PR for now. We're moving to a new parsing infrastructure which will likely require a different approach to support various flavors of javascript.

**Pull request closed without merging**